File: sql-chatbot-rails/lib/sql_chatbot/grammar/registry.rb
# frozen_string_literal: true

require "time"

module SqlChatbot
  module Grammar
    Entity = Struct.new(:name, :table, :display_label, :row_count, :primary_key,
                        :timestamps, :fields, :scopes, :associations, :ranking_candidates,
                        keyword_init: true) do
      def initialize(**kwargs)
        super(
          name: kwargs[:name],
          table: kwargs[:table],
          display_label: kwargs[:display_label] || kwargs[:name]&.capitalize,
          row_count: kwargs[:row_count] || 0,
          primary_key: kwargs[:primary_key] || "id",
          timestamps: kwargs[:timestamps] || {},
          fields: kwargs[:fields] || {},
          scopes: kwargs[:scopes] || {},
          associations: kwargs[:associations] || {},
          ranking_candidates: kwargs[:ranking_candidates] || []
        )
      end
    end

    Field = Struct.new(:column, :type, :nullable, :enum_values, :fk_to,
                       :user_facing_label, :searchable, keyword_init: true)

    Scope = Struct.new(:name, :where_clause, :param_slots, keyword_init: true)

    Association = Struct.new(:name, :kind, :target_entity, :join_clause,
                             :through_entity, keyword_init: true)

    class Registry
      attr_reader :entities, :aliases, :version, :generated_at, :framework

      def initialize(framework:, entities: {}, aliases: {})
        @framework = framework
        @entities = entities
        @aliases = aliases
        @version = 1
        @generated_at = Time.now.utc.iso8601
      end

      def find_entity(name)
        @entities[name.to_s]
      end

      def resolve_alias(term)
        return @aliases[term.to_s] if @aliases.key?(term.to_s)
        return term.to_s if @entities.key?(term.to_s)
        nil
      end
    end
  end
end

File: packages/agent/src/grammar/registry.ts
export interface Registry {
  entities: Record<string, Entity>;
  aliases: Record<string, string>;
  version: number;
  generatedAt: string;
  framework: 'rails' | 'django' | 'generic';
}

export interface Entity {
  name: string;
  table: string;
  displayLabel: string;
  rowCount: number;
  primaryKey: string;
  timestamps: { created?: string; updated?: string; deleted?: string };
  fields: Record<string, Field>;
  scopes: Record<string, Scope>;
  associations: Record<string, Association>;
  rankingCandidates: string[];
}

export type FieldType = 'int' | 'text' | 'bool' | 'timestamp' | 'decimal' | 'enum' | 'jsonb' | 'uuid';

export interface Field {
  column: string;
  type: FieldType;
  nullable: boolean;
  enumValues?: Record<string, number | string>;
  fkTo?: { entity: string; onColumn: string };
  userFacingLabel?: string;
  searchable: boolean;
}

export interface Scope {
  name: string;
  whereClause: string;
  paramSlots: string[];
}

export interface Association {
  name: string;
  kind: 'belongs_to' | 'has_many' | 'has_one' | 'has_many_through';
  targetEntity: string;
  joinClause: string;
  throughEntity?: string;
}

export function createRegistry(opts: { framework: Registry['framework'] }): Registry {
  return {
    entities: {},
    aliases: {},
    version: 1,
    generatedAt: new Date().toISOString(),
    framework: opts.framework,
  };
}

export function findEntity(r: Registry, name: string): Entity | null {
  return r.entities[name] ?? null;
}

export function resolveAlias(r: Registry, term: string): string | null {
  if (r.aliases[term]) return r.aliases[term];
  if (r.entities[term]) return term;
  return null;
}
